coralproject/talk
# [CORL-2672]: Refresh comments button on page visibility change

## What does this PR do?

This PR adds a `Refresh comments` button that shows whenever the page visibility changes (so, when someone returns to comments after locking then unlocking their phone or navigating to and back from another browser tab). Clicking the `Refresh comments` button refreshes the comments on the page. Clicking the `X` closes the button.

## These changes will impact:


## What changes to the GraphQL/Database Schema does this PR introduce?

It adds a `refreshStream` argument to the Story and Comment connections. This argument allows for comments to be refreshed fully, including all new edges that have come in since it was last refreshed.

## Does this PR introduce any new environment variables or feature flags?

n/a

## If any indexes were added, were they added to `INDEXES.md`?

n/a

## How do I test this PR?

Load the stream in the browser on a desktop/laptop. If you load up the all comments stream, you should not see a `Refresh comments` button. Switch to another browser tab and back. Or close your computer and reopen it to return to comments. You should now see the button show up. Click the button and it should refresh the comments. Test that it works and shows any new comments/replies that you have added (both before switching to focus on another tab and while switched and focusing on another tab).

Load the stream on a mobile device. Go to lock screen. Add comments to the stream in another window. Unlock your mobile device and return to the stream. See and click the `Refresh comments` button. See that it loads the stream and gets all new comments/replies. See that it goes away once you click to refresh. Test out clicking the `X` and seeing that the button disappears and doesn't refresh the stream. Ensure you can still comment/reply as usual and otherwise interact with the stream after refresh. See that there is an event sent when you click the button to refresh and when you x out the refresh button.

See that the `Refresh comments` button only shows when you are viewing Coral comments, and it is not visible when you are reading the story. It should be positioned at the top of the comments if you are at the top of Coral, and the Comments/Profile tabs and the post comment form are in view. As you scroll down, it should then be fixed at the top of your screen, visible as you view comments until you click to refresh/close the button.

Check that permalink conversation threads and the conversation threads in the admin moderate card modals are still working as expected.

Test that this works as expected with Q&A and says `Refresh questions` instead of `Refresh comments` in this case.

## Where any tests migrated to React Testing Library?

no

## How do we deploy this PR?

n/a
